### ISSUE TITLE
Fix failing unit tests

### ISSUE THREAD

**Issue by bokobza** [stratisproject/StratisBitcoinFullNode#214]
There are a few unit tests that are failing intermittently.
We need to fix them in order to get a clearer view of the sanity of our build and the code in general.

**Windows:**
RunWithDelayRunsTaskUntilCancelled
NotifyWithoutSyncFromRunsWithoutBroadcastingBlocks
StartLogsStartAndStop

**Linux:**
GetTimeOffsetReturnsCurrentUtcTimeOffset
StartWithDelayUsesIntervalForDelayRunsTaskUntilCancelled

**OS X:**
GetTimeOffsetReturnsCurrentUtcTimeOffset
RunWithDelayRunsTaskUntilCancelled
SpendableAmountConfirmedOnlyGivenBeingConfirmedAndSpentUnconfirmedReturnsAmount

In the attached file there is a bit more explanation on the failures.
[unittests.txt](https://github.com/stratisproject/StratisBitcoinFullNode/files/1131715/unittests.txt)

**Comment by Neurosploit** [comment]
Windows  10
While running all unit tests the tests in the TransactionModelsTest class fail as well as the LoadPoolTest_WithGoodTransactions test from the MempoolPersistenceTest class.'
Running them individually makes them pass.

See stacktraces below:

Test Name:	LoadPoolTest_WithGoodTransactions
Test FullName:	Stratis.Bitcoin.Tests.MemoryPool.MempoolPersistenceTest.LoadPoolTest_WithGoodTransactions
Test Source:	C:\StratisBitcoinFullNode\Stratis.Bitcoin.Tests\MemoryPool\MempoolPersistenceTest.cs : line 117
Test Outcome:	Failed
Test Duration:	0:00:00.128

Result StackTrace:	
at NBitcoin.BitcoinStream.ReadWriteBytes(Byte[]& data, Int32 offset, Int32 count)
   at NBitcoin.Protocol.VarString.ReadWrite(BitcoinStream stream)
   at NBitcoin.BitcoinStream.ReadWrite(Script data)
   at NBitcoin.BitcoinStream.ReadWrite(Script& script)
   at NBitcoin.TxIn.ReadWrite(BitcoinStream stream)
   at NBitcoin.BitcoinStream.ReadWrite[T](T& data)
   at NBitcoin.BitcoinStream.ReadWriteArray[T](T[]& data)
   at NBitcoin.BitcoinStream.ReadWriteList[TList,TItem](TList& data)
   at NBitcoin.Transaction.ReadWrite(BitcoinStream stream)
   at Stratis.Bitcoin.Tests.MemoryPool.MempoolPersistenceTest.<LoadPoolTest_WithGoodTransactions>d__8.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Tests\MemoryPool\MempoolPersistenceTest.cs:line 119
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Result Message:	System.IO.EndOfStreamException : No more byte to read


Test Name:	TransactionModelVerboseRenderNonCoinbaseTest
Test FullName:	Stratis.Bitcoin.Tests.RPC.Models.TransactionModelsTest.TransactionModelVerboseRenderNonCoinbaseTest
Test Source:	C:\StratisBitcoinFullNode\Stratis.Bitcoin.Tests\RPC\Models\TransactionModelsTest.cs : line 135
Test Outcome:	Failed
Test Duration:	0:00:00.118

Result StackTrace:	
at Stratis.Bitcoin.Tests.RPC.Models.TransactionModelsTest.TransactionModelVerboseRenderNonCoinbaseTest() in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Tests\RPC\Models\TransactionModelsTest.cs:line 137
----- Inner Stack Trace -----
   at NBitcoin.BitcoinStream.ReadWriteBytes(Byte[]& data, Int32 offset, Int32 count)
   at NBitcoin.Protocol.VarString.ReadWrite(BitcoinStream stream)
   at NBitcoin.BitcoinStream.ReadWrite(Script data)
   at NBitcoin.BitcoinStream.ReadWrite(Script& script)
   at NBitcoin.TxIn.ReadWrite(BitcoinStream stream)
   at NBitcoin.BitcoinStream.ReadWrite[T](T& data)
   at NBitcoin.BitcoinStream.ReadWriteArray[T](T[]& data)
   at NBitcoin.BitcoinStream.ReadWriteList[TList,TItem](TList& data)
   at NBitcoin.Transaction.ReadWrite(BitcoinStream stream)
   at NBitcoin.Transaction.Parse(String hex)
   at Stratis.Bitcoin.Tests.RPC.Models.TransactionModelsTest..cctor() in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Tests\RPC\Models\TransactionModelsTest.cs:line 27
Result Message:	
System.TypeInitializationException : The type initializer for 'Stratis.Bitcoin.Tests.RPC.Models.TransactionModelsTest' threw an exception.
---- System.IO.EndOfStreamException : No more byte to read

**Comment by Aprogiena** [comment]
I have this one failing for me, 2 times out of 3 tries:

**First run:**

Test Name:	WalletCanReorg
Test FullName:	Stratis.Bitcoin.IntegrationTests.WalletTests.WalletCanReorg
Test Source:	X:\Dev\Stratis\StratisBitcoinFullNode\Stratis.Bitcoin.IntegrationTests\WalletTests.cs : line 82
Test Outcome:	Failed
Test Duration:	0:00:44,801

Result StackTrace:	
at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Stratis.Bitcoin.IntegrationTests.TestHelper.WaitLoop(Func`1 act) in X:\Dev\Stratis\StratisBitcoinFullNode\Stratis.Bitcoin.IntegrationTests\TestHelper.cs:line 16
   at Stratis.Bitcoin.IntegrationTests.WalletTests.WalletCanReorg() in X:\Dev\Stratis\StratisBitcoinFullNode\Stratis.Bitcoin.IntegrationTests\WalletTests.cs:line 203
Result Message:	System.OperationCanceledException : The operation was canceled.


**Second run:**
Test Name:	WalletCanReorg
Test FullName:	Stratis.Bitcoin.IntegrationTests.WalletTests.WalletCanReorg
Test Source:	X:\Dev\Stratis\StratisBitcoinFullNode\Stratis.Bitcoin.IntegrationTests\WalletTests.cs : line 82
Test Outcome:	Failed
Test Duration:	0:02:01,901

Result StackTrace:	
at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Stratis.Bitcoin.IntegrationTests.TestHelper.WaitLoop(Func`1 act) in X:\Dev\Stratis\StratisBitcoinFullNode\Stratis.Bitcoin.IntegrationTests\TestHelper.cs:line 16
   at Stratis.Bitcoin.IntegrationTests.WalletTests.WalletCanReorg() in X:\Dev\Stratis\StratisBitcoinFullNode\Stratis.Bitcoin.IntegrationTests\WalletTests.cs:line 224
Result Message:	System.OperationCanceledException : The operation was canceled.

**Comment by quantumagi** [comment]
Any test that is timing sensitive will always fail sooner or later. I suggest that we somehow remove these tests from automated testing so as not to interfere with pull requests.

**Comment by Aprogiena** [comment]
@quantumagi : I did not check the code, so I'm not sure why those tests are failing. Could you please briefly describe what the problem is? I'd rather explore the problem and fix it than just remove those tests completely.

**Comment by quantumagi** [comment]
@Aprogiena: The code is setting up asynchronous loops and cancellation tokens with intervals that can't be strictly enforced as exact execution times would depend on work load. As such tests that count how often the code is invoked over a particular time interval can't deliver fully deterministic results.

**Comment by Aprogiena** [comment]
I'm not sure I understand what you mean by that. Certain things may definitely depend on workload. That's OK, but it is unlikely for a test to do so much work that it could not be possible to think off a hard limit that will always be sufficient should the test pass. 

Let's dig deeper into this and more specifically:
I have that WalletCanReorg test that fails for me, now it is at WalletTests.cs:line 231:

`TestHelper.WaitLoop(() => TestHelper.AreNodesSynced(stratisReceiver, stratisSender));`

And during the execution of this test, CPU is unused, so this is not due to workload, but rather due to some kind of timers somewhere that we need to be triggered.

Now, checking WaitLoop gives me info that there is max 30 second wait for the condition to be satisfied. And the same WaitLoop is used for many other things in this test. The question is - is 30 seconds enough for nodes to get synced there? I'm not familiar with that test or the internals of the node syncing, but obviously it might not be the case. So the author of the code and the test should know what is the upper limit on this operation and should specify this limit rather than 30 second limit.

So I tried to put there 2 minutes WaitLoop. That failed as well (on different line of the test, however). So I put 5 minutes there. It is still failing on the last reorg (line 252). This suggests to me that the test is poorly written, the author should know how long this could take at max. If there were good logs, one could see what is happening there.

Some kind of limitation on how much time a single test can consume might be good. In previous project we had categories of tests and each category could consume some maximum amount of time.

To summarize - if it depends on workload, enough time should be given for the test and this should ensure that the test always passes. If it depends on some events that needs to trigger (like timers), it might be possible to add code that triggers them on demand - for the purpose of testing, or again, we can give the test enough time to complete.

As for the manual triggering - one can imagine a cleanup task to be done every couple of hours and you don't want to have a test that does nothing but waits for the timer to trigger. So in that case you should implement a way to manually trigger the timer event. This should be done probably for anything that needs more than X (= 5?) minutes.

**Comment by quantumagi** [comment]
Yes I have tried increasing the time granularity by a factor of 10 to see if I could do something to make those tests work more consistently. However I only had limited success. Increasing the times by another factor of 10 may help again, though perhaps not perfectly. At some points these tests will start running for a significant time, although I would not discourage you to try it. I am just left with the feeling that on some odd occasion they may fail again. Somewhat of a nuisance that is. By the way, I have observed significantly more issues with the release version vs the debug code. Maybe that is something worth exploring further.

**Comment by bokobza** [comment]
More failing tests in Release. It seems to be NBitcoin.Parse that's causing trouble.
https://ci.appveyor.com/project/stratis/stratisbitcoinfullnode/build/1.0.2.29/job/3sh14b5a6hk38pb8/tests
![image](https://user-images.githubusercontent.com/1867877/28335382-186066c6-6bf6-11e7-8f16-206903443a36.png)

**Comment by bokobza** [comment]
https://github.com/MetacoSA/NBitcoin/commit/5bf64a0ccc0a2df6479da0527e09d834499d0745

**Comment by bokobza** [comment]
All the issues here have been resolved so I'm closing this.
